File: app/src/test/java/com/example/recipeclipper/fake/FakeTimerAlarmScheduler.kt
package com.example.recipeclipper.fake

import com.example.recipeclipper.data.TimerAlarmScheduler
import com.example.recipeclipper.data.model.StepAlarm

/**
 * Keeps the alarms as real state (recipe id and step to the alarm), the way `AlarmManager`
 * does, so a test can assert what is pending after a sequence of calls; every call is also
 * recorded in order.
 */
class FakeTimerAlarmScheduler : TimerAlarmScheduler {

    val pending = mutableMapOf<Pair<Long, Int>, StepAlarm>()
    val scheduleCalls = mutableListOf<StepAlarm>()
    val cancelCalls = mutableListOf<Pair<Long, Int>>()

    override fun schedule(alarm: StepAlarm) {
        scheduleCalls += alarm
        pending[alarm.recipeId to alarm.step] = alarm
    }

    override fun cancel(recipeId: Long, step: Int) {
        cancelCalls += recipeId to step
        pending.remove(recipeId to step)
    }
}

File: ios/RecipeClipperTests/Fakes/FakeTimerAlarmScheduler.swift
import Foundation
@testable import RecipeClipper

/// Keeps the pending alerts as real state (recipe id and step to the alarm), as the
/// notification center does, and records every call in order.
final class FakeTimerAlarmScheduler: TimerAlarmScheduler {
    struct Key: Hashable {
        let recipeId: Int64
        let step: Int
    }

    private(set) var pending: [Key: StepAlarm] = [:]
    private(set) var scheduleCalls: [StepAlarm] = []
    private(set) var cancelCalls: [Key] = []
    private(set) var replaceAllCalls: [(recipeId: Int64, alarms: [StepAlarm])] = []

    func schedule(_ alarm: StepAlarm) {
        scheduleCalls.append(alarm)
        pending[Key(recipeId: alarm.recipeId, step: alarm.step)] = alarm
    }

    func cancel(recipeId: Int64, step: Int) {
        cancelCalls.append(Key(recipeId: recipeId, step: step))
        pending[Key(recipeId: recipeId, step: step)] = nil
    }

    func replaceAll(recipeId: Int64, with alarms: [StepAlarm]) {
        replaceAllCalls.append((recipeId, alarms))
        pending = pending.filter { $0.key.recipeId != recipeId }
        for alarm in alarms { pending[Key(recipeId: alarm.recipeId, step: alarm.step)] = alarm }
    }
}
